Add MeshInstance3D primitive conversion options

kitbdev opened the pull request:
- closes https://github.com/godotengine/godot-proposals/issues/9685
- closes https://github.com/godotengine/godot-proposals/issues/10802
- closes https://github.com/godotengine/godot-proposals/issues/5868

Adds Bounding Box, Capsule, Cylinder, Sphere, and Primitive options. 

![image](https://github.com/user-attachments/assets/7d7ba100-21c8-49e4-badf-839bf6aa6d9f)

The primitive option is only enabled when a supported PrimitiveMesh is selected. This will make the corresponding collision shape. When multiple meshes are selected, each PrimitiveMesh will create their own shape and all others will do nothing.
The other options first check for their corresponding PrimitiveMesh, and if it is not found the mesh AABB is used instead.
I didn't call the Bounding Box option AABB since we don't have an AABB collision shape type. It uses a box and it may be rotated so I thought it could be misleading.
I made sure that the collision shape node will be offset to match the AABB.
The sphere radius is based on the longest axis. 

When Capsule or Cylinder is selected an axis option will show, to allow you to configure the long axis of it. The longest axis is the default. I could add a shortest axis option if it is wanted.

![image](https://github.com/user-attachments/assets/70d1b3f4-acb6-416d-b139-987b94e0962a)

kitbdev committed:
Add MeshInstance3D primitive conversion options

kitbdev commented on editor/plugins/mesh_instance_3d_editor_plugin.cpp:
```diff
@@ -103,31 +107,249 @@ Vector<Ref<Shape3D>> MeshInstance3DEditor::create_shape_from_mesh(Ref<Mesh> p_me
+		Ref<Mesh> m = instance->get_mesh();
+		if (m.is_null()) {
+			continue;
+		}
+		if (((Ref<BoxMesh>)m).is_valid() || ((Ref<CapsuleMesh>)m).is_valid() || ((Ref<CylinderMesh>)m).is_valid() || ((Ref<SphereMesh>)m).is_valid()) {
+			disable_primitive = false;
+			break;
+		}
+	}
+	if (disable_primitive && shape_type->get_selected() == SHAPE_TYPE_PRIMITIVE) {
+		shape_type->select(SHAPE_TYPE_TRIMESH);
+	}
+
+	shape_type->set_item_disabled(shape_type->get_popup()->get_item_index(SHAPE_TYPE_PRIMITIVE), disable_primitive);
+
+	ShapeType selected_shape_type = (ShapeType)shape_type->get_selected_id();
+	bool shape_axis_visible = selected_shape_type == SHAPE_TYPE_CAPSULE || selected_shape_type == SHAPE_TYPE_CYLINDER;
+	shape_axis->set_visible(shape_axis_visible);
+	shape_axis_label->set_visible(shape_axis_visible);
+}
+
+void MeshInstance3DEditor::_shape_type_selected(int p_option) {
+	bool shape_axis_visible = (ShapeType)p_option == SHAPE_TYPE_CAPSULE || (ShapeType)p_option == SHAPE_TYPE_CYLINDER;
+	shape_axis->set_visible(shape_axis_visible);
+	shape_axis_label->set_visible(shape_axis_visible);
+}
+
 void MeshInstance3DEditor::_create_collision_shape() {
 	int placement_option = shape_placement->get_selected();
 	int shape_type_option = shape_type->get_selected();
 
 	EditorSelection *editor_selection = EditorNode::get_singleton()->get_editor_selection();
 	EditorUndoRedoManager *ur = EditorUndoRedoManager::get_singleton();
 
+	String placement_action_name = placement_option == SHAPE_PLACEMENT_SIBLING ? TTR("Create %s Collision Shape Sibling") : TTR("Create %s Static Body Child");
+
 	switch (shape_type_option) {
 		case SHAPE_TYPE_TRIMESH: {
-			ur->create_action(TTR(placement_option == SHAPE_PLACEMENT_SIBLING ? "Create Trimesh Collision Shape Sibling" : "Create Trimesh Static Body"));
+			ur->create_action(vformat(placement_action_name, TTR("Trimesh")));
```
I believe these were incorrect for translations, so I changed how they use TTR.
It should now also be easier to extend if more placement options are added.

Calinou replied:
Indeed, `TTR()` should not contain conditions within it (the condition should be outside).

Calinou commented on editor/plugins/mesh_instance_3d_editor_plugin.cpp:
```diff
@@ -630,12 +855,39 @@ MeshInstance3DEditor::MeshInstance3DEditor() {
+
+	shape_axis = memnew(OptionButton);
+	shape_axis->set_h_size_flags(SIZE_EXPAND_FILL);
+	shape_axis->add_item(TTR("Longest axis"), SHAPE_AXIS_LONGEST);
+	shape_axis->set_item_tooltip(-1, TTR("Create the shape along the longest axis of the mesh's AABB."));
+	shape_axis->add_item(TTR("X axis"), SHAPE_AXIS_X);
+	shape_axis->set_item_tooltip(-1, TTR("Create the shape along the local X axis."));
+	shape_axis->add_item(TTR("Y axis"), SHAPE_AXIS_Y);
+	shape_axis->set_item_tooltip(-1, TTR("Create the shape along the local Y axis."));
+	shape_axis->add_item(TTR("Z axis"), SHAPE_AXIS_Z);
+	shape_axis->set_item_tooltip(-1, TTR("Create the shape along the local Z axis."));
```
Use Title Case for option names for consistency with the rest of the editor:

```suggestion
	shape_axis->add_item(TTR("Longest Axis"), SHAPE_AXIS_LONGEST);
	shape_axis->set_item_tooltip(-1, TTR("Create the shape along the longest axis of the mesh's AABB."));
	shape_axis->add_item(TTR("X Axis"), SHAPE_AXIS_X);
	shape_axis->set_item_tooltip(-1, TTR("Create the shape along the local X axis."));
	shape_axis->add_item(TTR("Y Axis"), SHAPE_AXIS_Y);
	shape_axis->set_item_tooltip(-1, TTR("Create the shape along the local Y axis."));
	shape_axis->add_item(TTR("Z Axis"), SHAPE_AXIS_Z);
	shape_axis->set_item_tooltip(-1, TTR("Create the shape along the local Z axis."));
```

kitbdev replied:
Looking through some code, I think `X-Axis` is more preferred.

Calinou reviewed (commented):
Tested locally, it works as expected. The feature makes sense to me.

However, there's an issue if you try using this with very thin models (e.g. QuadMesh or PlaneMesh). The Bounding Box mode will create a 0.0-units-thick shape that won't collide with anything else when using Jolt. We can avoid this by ensuring each dimension of the bounding box is at least `0.001` units thick (same for cylinders). This issue doesn't occur in GodotPhysics3D.

**Edit:** I've reported this in a dedicated issue: https://github.com/godotengine/godot/issues/101893

kitbdev commented:
It seems like 0 thickness shapes are supported now, but they aren't recommended. So I guess I should still use `0.001` as a minimum.

kitbdev commented:
Added size minimum for box and cylinder height.
Added size minimum for sphere and capsule radius, because when on Jolt, an error was shown if it was 0. It didn't seem to be needed for cylinder radius, since it didn't go to a size of 0.

Calinou reviewed (approved):
Tested on all mesh types with default parameters. Works great now :slightly_smiling_face: 

Code looks good to me.

kitbdev commented:
Does this need more reviews or is it ready to merge?

Repiteo commented:
Thanks!
